File: apple_music_api/AppleMusicApi.py
import datetime
import json
import jwt
import requests

class AppleMusicApi:
    def __init__(self, key_id, issuer, key_filename):
        key_file = open(key_filename, 'r')
        secret_key = key_file.read()
        key_file.close()
        iat = int(datetime.datetime.utcnow().strftime("%s"))
        exp = int((datetime.datetime.utcnow() + datetime.timedelta(hours=12)).strftime("%s"))
        algorithm = 'ES256'

        payload = {
            'iss': issuer,
            'iat': iat,
            'exp': exp
        }

        headers = {
            'alg': algorithm,
            'kid': key_id
        }

        token_bytes = jwt.encode(payload, secret_key, algorithm=algorithm, headers=headers)
        self.developer_token = token_bytes.decode()

    def _get(self, url, params={}):
        headers = { "Authorization": "Bearer {0}".format(self.developer_token) }
        response = requests.get(url, params=params, headers=headers)

        return json.loads(response.text)

    def get_album(self, album_id, storefront='us'):
        url = "https://api.music.apple.com/v1/catalog/{0}/albums/{1}".format(storefront, album_id)
        album = self._get(url)

        return album

    def get_track(self, track_id, storefront='us'):
        url = "https://api.music.apple.com/v1/catalog/{0}/songs/{1}".format(storefront, track_id)
        track = self._get(url)

        return track

    def search(self, term, limit=25, offset=0, types='artists, albums, songs, playlists, stations', storefront='us'):
        url = "https://api.music.apple.com/v1/catalog/{0}/search".format(storefront)
        params = {
            'term': term,
            'limit': limit,
            'offset': offset,
            'types': types
        }
        results = self._get(url, params)

        return results

File: lib/apple_music_api/apple_music_api.rb
# frozen_string_literal: true
# typed: strict

class AppleMusicAPI
  include Singleton

  sig {void}
  def initialize
    issuer = ENV['APPLE_KEY_ISSUER']
    iat = Time.now.to_i
    exp = 12.hours.from_now.to_i
    payload = {
      'iss': issuer,
      'iat': iat,
      'exp': exp,
    }

    key_id = ENV['APPLE_KEY_ID']
    headers = {
      'alg': 'ES256',
      'kid': key_id
    }
    secret_key = OpenSSL::PKey::EC.new(ENV['APPLE_KEY'])

    @developer_token = T.let(
      JWT.encode(payload, secret_key, 'ES256', headers),
      String,
    )
  end

  sig do
    params(
      path: String,
      query: T::Hash[Symbol, T.untyped]
    )
    .returns(
      T::Hash[String, T.untyped]
    )
  end
  private def _get(path, query={})
    response = Excon.get(
      "https://api.music.apple.com/v1/catalog/us/#{path}",
      query: query,
      headers: {"Authorization": "Bearer #{@developer_token}"},
    )

    JSON.parse(response.body)
  end

  sig {params(album_id: String).returns(T::Hash[String, T.untyped])}
  def get_album(album_id)
    response = _get("albums/#{album_id}")

    response["data"].first
  end

  sig {params(track_id: String).returns(T::Hash[String, T.untyped])}
  def get_track(track_id)
    response = _get("songs/#{track_id}")

    response["data"].first
  end

  sig do
    params(
      title: String,
      artist_name: String
    )
    .returns(
      T.nilable(T::Hash[String, T.untyped])
    )
  end
  def search_album(title, artist_name)
    params = {
      term: "#{title} #{artist_name}".delete(','),
      limit: 1,
      offset: 0,
      types: 'albums',
    }
    response = _get("search", params)
    results = response.dig("results", "albums", "data") || []

    results.first
  end

  sig do
    params(
      title: String,
      artist_name: String
    )
    .returns(
      T.nilable(T::Hash[String, T.untyped])
    )
  end
  def search_track(title, artist_name)
    params = {
      term: "#{title} #{artist_name}".delete(','),
      limit: 1,
      offset: 0,
      types: 'songs',
    }
    response = _get("search", params)
    results = response.dig("results", "songs", "data") || []

    results.first
  end
end
